src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp

```diff
@@ -0,0 +1,488 @@
+        spacetime_unit_normal_vector,
+    const gsl::not_null<tnsr::a<DataVector, Dim, Frame::Inertial>*>
+        incoming_null_one_form,
+    const gsl::not_null<tnsr::a<DataVector, Dim, Frame::Inertial>*>
+        outgoing_null_one_form,
+    const gsl::not_null<tnsr::A<DataVector, Dim, Frame::Inertial>*>
+        incoming_null_vector,
+    const gsl::not_null<tnsr::A<DataVector, Dim, Frame::Inertial>*>
+        outgoing_null_vector,
+    const gsl::not_null<tnsr::aa<DataVector, Dim, Frame::Inertial>*>
+        projection_ab,
+    const gsl::not_null<tnsr::Ab<DataVector, Dim, Frame::Inertial>*>
+        projection_Ab,
+    const gsl::not_null<tnsr::AA<DataVector, Dim, Frame::Inertial>*>
+        projection_AB,
+    const gsl::not_null<tnsr::aa<DataVector, Dim, Frame::Inertial>*>
+        char_projected_rhs_dt_v_psi,
+    const gsl::not_null<tnsr::iaa<DataVector, Dim, Frame::Inertial>*>
+        char_projected_rhs_dt_v_zero,
+    const gsl::not_null<tnsr::aa<DataVector, Dim, Frame::Inertial>*>
+        char_projected_rhs_dt_v_minus,
+    const gsl::not_null<tnsr::a<DataVector, Dim, Frame::Inertial>*>
+        constraint_char_zero_plus,
+    const gsl::not_null<tnsr::a<DataVector, Dim, Frame::Inertial>*>
+        constraint_char_zero_minus,
+    const gsl::not_null<std::array<DataVector, 4>*> char_speeds,
+
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /* face_mesh_velocity */,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& pi,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& phi,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& spacetime_metric,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& /* coords */,
+    const Scalar<DataVector>& gamma1, const Scalar<DataVector>& gamma2,
+    const Scalar<DataVector>& lapse,
+    const tnsr::a<DataVector, Dim, Frame::Inertial>& gauge_source,
+    const tnsr::ab<DataVector, Dim, Frame::Inertial>&
+        spacetime_deriv_gauge_source,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& dt_pi,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& dt_phi,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& dt_spacetime_metric,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& d_pi,
+    const tnsr::ijaa<DataVector, Dim, Frame::Inertial>& d_phi,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& d_spacetime_metric)
+    const noexcept {
+  TempBuffer<tmpl::list<::Tags::TempScalar<0, DataVector>,
+                        ::Tags::TempI<0, Dim, Frame::Inertial, DataVector>,
+                        ::Tags::Tempii<0, Dim, Frame::Inertial, DataVector>,
+                        ::Tags::Tempa<0, Dim, Frame::Inertial, DataVector>,
+                        ::Tags::Tempia<0, Dim, Frame::Inertial, DataVector>>>
+      local_buffer(get_size(get<0>(normal_covector)));
+
+  //   auto& lapse = get<::Tags::TempScalar<0, DataVector>>(local_buffer);
+  auto& shift =
```

Comment:
why are you computing these instead of grabbing them from the temporary tags? lapse, shift, spatial metric, spacetime normals, inverse spatial metric, inverse spacetime metric, three index constraint, are all computed in the bulk and could be sliced. Naively I'd expect that, at least things like the inverse spacetime metric and three-index constraint would be better to slice than to recompute. The right thing to do is whatever is fastest, but it's not completely obvious to me in general which is faster, though I assume slicing will win for all except maybe the most trivial tags. If you'd like to defer this to a followup PR, that's fine, but a code comment and an issue would be good

Reply:
ping

Reply:
I wasn't sure of which would be faster either. There would be the computation cost and the memory allocation cost. I thought at least in the case when these vars are being recomputed here, the memory allocation can be done in one buffer, while the same won't be possible when slicing from volume. But not having profiled this, I am quite happy to change the behavior here. Pushed a patch that makes use of as many temp tags as available.

Reply:
There's no memory allocation cost for the slicing, it's already paid for just by having an interface. We have one Variables that stores everything on the interface, so switching actually means fewer allocations :)